ait/gui/__init__.py

```diff
@@ -985,6 +993,178 @@ def handle():
+            ]
+    """
+    global playback
+    ranges = []
+
+    # Loop through each packet from database
+    packets = list(playback.dbconn.query('SHOW MEASUREMENTS').get_points())
+    for i in range(len(packets)):
+
+        # Add packet name
+        packet_name = packets[i]['name']
+        ranges.append([packet_name])
+
+        # Add start time and end time
+        point_query = 'SELECT * FROM "{}"'.format(packet_name)
+        points = list(playback.dbconn.query(point_query).get_points())
+        # Round start time down to nearest second
+        start_time = points[0]['time'][:19] + 'Z'
+        ranges[i].append(start_time)
+        # Round end time up to nearest second
+        end_time = points[len(points) - 1]['time']
+        dt = datetime.strptime(end_time, '%Y-%m-%dT%H:%M:%S.%fZ')
+        if dt.microsecond != 0:
+            dt += timedelta(0, 1)
+        end_time = dt.strftime('%Y-%m-%dT%H:%M:%SZ')
+        ranges[i].append(end_time)
+
+    return json.dumps(ranges)
+
+
+@App.route('/playback/query', method='POST')
+def handle():
+    """Set playback query with packet name, start time, and end time from form"""
+    global playback
+    tlm_dict = tlm.getDefaultDict()
+
+    # Get values from form
+    packet = bottle.request.forms.get('packet')
+    start_time = bottle.request.forms.get('startTime')
+    end_time = bottle.request.forms.get('endTime')
+    uid = tlm_dict[packet].uid
+
+    # Query packet and time range from database
+    point_query = 'SELECT * FROM "{}" WHERE time >= \'{}\' AND time <= \'{}\''.format(packet, start_time, end_time)
+    points = list(playback.dbconn.query(point_query).get_points())
+
+    # Sort data of query according to fields in tlm dictionary
+    pkt = tlm_dict[packet]
+    fields = pkt.fields
+    field_names = []
+    for i in range(len(fields)):
+        field_names.append(fields[i].name)
+    # Put query into a map of {timestamp: data}
+    for i in range(len(points)):
+        # Round time down to nearest 0.1 second
+        timestamp = str(points[i]['time'][:21] + 'Z')
+        data = bytearray(1) + struct.pack('>I', uid)
+        for j in range(len(field_names)):
+            data += struct.pack('>H', points[i][field_names[j]])
+        if playback.query.has_key(timestamp):
+            playback.query[timestamp].append(data)
+        else:
+            playback.query[timestamp] = [data]
+
+
+@App.route('/playback/send', method='POST')
+def handle():
+    """Send timestamp to be put into playback queue if in database"""
+    global playback
+    timestamp = bottle.request.forms.get('timestamp')
+    if playback.query.has_key(timestamp):
+        playback.queue.append(timestamp)
+
+
+@App.route('/playback/playback')
+def handle():
+    """Return historical telemetry packets as if it were realtime to client"""
+    global playback
+    global realtime_wsock
+
+    # Close realtime socket
+    realtime_wsock.closed = True
```

Comment:
It seems like it would be easier and cleaner to track whether we're displaying real-time or historical data and gate telemetry flow some other way instead of forcing connections to be opened and closed. I'd rather see state tracked and update the logic elsewhere (such as just prior to emitting the telemetry to the websocket). Thoughts?

Reply:
That would definitely be cleaner. I just updated it!